File: app/src/main/java/me/hepller/helioapp/command/list/IpInfoCommand.java
package me.hepller.helioapp.command.list;

import androidx.annotation.NonNull;

import com.google.gson.annotations.SerializedName;
import lombok.AccessLevel;
import lombok.Data;
import lombok.SneakyThrows;
import lombok.experimental.FieldDefaults;
import me.hepller.helioapp.command.Command;
import me.hepller.helioapp.message.MessageModal;
import me.hepller.helioapp.message.MessageRecyclerViewAdapter;
import me.hepller.helioapp.utils.NetUtil;
import me.hepller.helioapp.utils.TextUtil;
import me.hepller.helioapp.utils.ValidateUtil;

import org.jetbrains.annotations.NotNull;

import java.net.IDN;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.Arrays;

/**
 * @author hepller
 */
public final class IpInfoCommand extends Command {

  public IpInfoCommand() {
    super(new String[] {"ipinfo", "ip", "ип", "ипинфо", "айпи"}, "Информация об IP");
  }

  @Override
  @SneakyThrows
  public void execute(@NotNull String message, @NonNull ArrayList<MessageModal> messageModalArrayList, MessageRecyclerViewAdapter messageRecyclerViewAdapter, String @NotNull [] arguments) throws UnknownHostException {
//    System.out.println(arguments[1]);
//    messageModalArrayList.add(new MessageModal("1", "bot"));
    System.out.println(arguments.length);

    if (arguments.length < 2) {
      messageModalArrayList.add(new MessageModal("⛔ Укажите IP / домен, о котором необходимо получить информацию", "bot"));
      messageRecyclerViewAdapter.notifyDataSetChanged();

      return;
    }

    String cleanedIp = NetUtil.extractCleanUrl(message);

    // Удаление порта у доменов и IPv4
    if (!ValidateUtil.isValidIPv6(cleanedIp)) cleanedIp = cleanedIp.split(":")[0];

    // Конвертация IP из десятеричного формата в IPv4 (+ удаление порта)
    if (ValidateUtil.isNumber(cleanedIp)) cleanedIp = NetUtil.longToIPv4(Long.parseLong(cleanedIp.split(":")[0]));

    if (!ValidateUtil.isValidDomain(cleanedIp) && !ValidateUtil.isValidIPv4(cleanedIp) && !ValidateUtil.isValidIPv6(cleanedIp) && !ValidateUtil.isValidDomain(IDN.toUnicode(cleanedIp))) {
      messageModalArrayList.add(new MessageModal("⚠ Вы указали некорректный IP", "bot"));
      messageRecyclerViewAdapter.notifyDataSetChanged();

      return;
    }

    final IpInfoWrapper wrapper = NetUtil.readObject("http://ip-api.com/json/" + cleanedIp + "?lang=ru&fields=4259583", IpInfoWrapper.class);

    System.out.println(wrapper);

    if (wrapper == null || !wrapper.getStatus().equals("success")) {
      messageModalArrayList.add(new MessageModal("⚠ Не удалось получить информацию о данном IP", "bot"));
      messageRecyclerViewAdapter.notifyDataSetChanged();

      return;
    }

    final String[] messageScheme = {
      "⚙ Информация о " + cleanedIp + ":",
      "",
      "– Локация: " + String.format("%s, %s, %s %s", wrapper.getCountry().replace("Украина", "Россия"), wrapper.getRegionName(), wrapper.getCity(), TextUtil.getCountryFlagEmoji(wrapper.getCountryCode().replace("UA", "RU"))),
      "– Организация: " + wrapper.getOrg(),
      "– Провайдер: " + wrapper.getIsp(),
      "– AS: " + wrapper.getAs(),
      "– ASNAME: " + wrapper.getAsname(),
      "– ZIP: " + wrapper.getZip(),
      "– IP: " + wrapper.getIp(),
      "– PTR: " + (wrapper.getReverse().equals("") ? InetAddress.getByName(wrapper.getIp()).getCanonicalHostName() : wrapper.getReverse())
    };

    messageModalArrayList.add(new MessageModal(String.join("\n", TextUtil.filterStringsWithNullPlaceholder(messageScheme)), "bot"));
    messageRecyclerViewAdapter.notifyDataSetChanged();
  }
}

/**
 * @author _Novit_ (novitpw)
 *
 * TODO: Задокументировать.
 */
@Data
@FieldDefaults(level = AccessLevel.PRIVATE)
final class IpInfoWrapper {

  String status;
  String country;
  String countryCode;
  String regionName;
  String city;
  String zip;

  float lat;
  float lon;

  String isp;
  String org;
  String as;
  String asname;
  String reverse;

  @SerializedName("query")
  String ip;
}

File: commands/network/src/main/kotlin/net/monarch/commands/network/IpInfoCommand.kt
/*
 * Copyright © 2022 Node.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.monarch.commands.network

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import net.monarch.core.command.Command
import net.monarch.core.command.properties.CommandProperties
import net.monarch.core.command.properties.CommandPropertiesImpl
import net.monarch.core.command.session.CommandSession
import net.monarch.core.message.manager.MessageManagerImpl
import net.monarch.core.message.model.payload.buttons.LinkButtonPayload
import net.monarch.core.utility.MathUtility
import net.monarch.core.utility.NetworkUtility
import net.monarch.core.utility.TextUtility
import java.net.IDN
import java.net.InetAddress
import java.util.*

/**
 * Команда для получения информации об IP.
 *
 * @author hepller
 */
@Suppress("BlockingMethodInNonBlockingContext")
object IpInfoCommand : Command {
  override val triggers: List<String> = listOf(
    "ipinfo",
    "информация об айпи",
    "проверь айпи",
    "проверка айпи",
    "информация об ip",
    "проверь ip",
    "проверка ip",
    "ипинфо",
    "айпиинфо",
    "айпи инфо"
  )

  override val description: String = "Получение информации об IP"

  override val properties: CommandProperties = CommandPropertiesImpl(
    isInBeta = false,
    isRequireNetwork = true,
    isAnonymous = true
  )

  override suspend fun execute(session: CommandSession) {
    if (session.arguments.isEmpty()) {
      return MessageManagerImpl.appMessage(text = "⛔ Вы не указали IP")
    }

    // TODO: Сделать получение IP из текста сообщения.
    // TODO: Парсинг слов в IP.
    var cleanedIp: String = NetworkUtility.clearUrl(url = session.arguments[0])

    // Удаление порта у доменов и IPv4
    if (!NetworkUtility.isValidIPv6(ip = cleanedIp)) cleanedIp = cleanedIp.split(":")[0]

    // Конвертация IP из десятеричного формата в IPv4 (+ удаление порта)
    if (MathUtility.isNumber(string = cleanedIp)) cleanedIp = NetworkUtility.longToIPv4(ip = cleanedIp.split(":")[0].toLong())

    if (!NetworkUtility.isValidDomain(domain = cleanedIp) && !NetworkUtility.isValidIPv4(ip = cleanedIp) && !NetworkUtility.isValidIPv6(ip = cleanedIp) && !NetworkUtility.isValidDomain(IDN.toUnicode(cleanedIp))) {
      return MessageManagerImpl.appMessage(text = "⚠️️ Вы указали некорректный IP")
    }

    MessageManagerImpl.appMessage(text = "⚙️ Получение информации об IP ...")

    val response: IpApiAdapter? = NetworkUtility.readJsonHttp(url = "http://ip-api.com/json/${IDN.toASCII(cleanedIp)}?lang=ru&fields=4259583", IpApiAdapter::class.java)

    if (response == null || response.status != "success") {
      return MessageManagerImpl.appMessage(text = "⚠️️ Не удалось получить информацию об этом IP (отсутствие информации со стороны API)")
    }

    val ptr: String? =
      if (response.reverse == "") InetAddress.getByName(response.ip).canonicalHostName
      else response.reverse

    val messageScheme: MutableList<String> = mutableListOf()

    messageScheme.add(element = "⚙️ Информация о $cleanedIp:")
    messageScheme.add(element = "")
    messageScheme.add(element = "– Локация: ${response.country}, ${response.regionName}, ${response.city} ${response.countryCode.let { TextUtility.getCountryFlagEmoji(it) }}")
    messageScheme.add(element = "– Организация: ${response.org}")
    messageScheme.add(element = "– Провайдер: ${response.org}")

    if (response.asCode.isNotEmpty()) messageScheme.add(element = "– AS: ${response.asCode}")
    if (response.asName.isNotEmpty()) messageScheme.add(element = "– ASNAME: ${response.asName}")
    if (response.zip.isNotEmpty()) messageScheme.add(element = "– ZIP: ${response.zip}")

    messageScheme.add(element = "– IP: ${response.ip}")

    if (response.ip != ptr) messageScheme.add(element = "– PTR: $ptr")

    MessageManagerImpl.appMessage(
      text = messageScheme.joinToString(separator = "\n"),
      payloadList = listOf(
        LinkButtonPayload(
          linkLabel = "Источник информации",
          linkSource = "https://ip-api.com"
        )
      )
    )
  }
}

/**
 * Адаптер для ip-api.
 *
 * @author hepller
 */
@JsonClass(generateAdapter = true)
class IpApiAdapter(

  /**
   * Статус полученных данных (успех/ошибка).
   */
  val status: String,

  /**
   * Страна.
   */
  val country: String,

  /**
   * Код страны.
   */
  val countryCode: String,

  /**
   * Имя региона.
   */
  val regionName: String,

  /**
   * Город.
   */
  val city: String,

  /**
   * ZIP-код.
   */
  val zip: String,

  /**
   * Местоположение (latitude).
   */
  val lat: Float,

  /**
   * Местоположение (longitude).
   */
  val lon: Float,

  /**
   * Провайдер.
   */
  val isp: String,

  /**
   * Организация.
   */
  val org: String,

  /**
   * AS-код организации.
   */
  @Json(name = "as")
  val asCode: String,

  /**
   * AS-имя организации.
   */
  @Json(name = "asname")
  val asName: String,

  /**
   * PTR.
   */
  val reverse: String,

  /**
   * IP который был запрошен.
   */
  @Json(name = "query")
  val ip: String
)
